tests/apps/tests.py

```diff
@@ -49,6 +53,34 @@ def test_ready(self):
         # Non-master app registries are populated in __init__.
         self.assertTrue(Apps().ready)
 
+    def test_populate_reentrance(self):
+        """
+        Running django.setup() during an ongoing setup() results in a deadlock.
+
+        To prevent Django's tests from hanging indefinitely if a deadlock
```

Comment:
I'm not sure if this is working properly. I reverted django/ and the test seems to hang.